### ISSUE TITLE
allow configuring where the accessible terminal view reopens to

### ISSUE THREAD

**Issue by origine999** [microsoft/vscode#194547]
Type: <b>Bug</b>

There is an accessibility regression from 1.82 > 1.83 with the integrated terminal.

In 1.83, with JAWS
1. Run a command in integrated terminal
2. Press Shift+Tab

Actual: JAWS is silent. Experience is clunky, up/down seems to write commands in the terminal input. Much worse experience versus 1.82
Expected: Focus moves to the readonly edit where I can review output.

Note: Yes, Alt+F2 is a nice buffer, but Shift+Tab should take you to an accessible readonly edit like it did in 1.82

In 1.82, that would be:
1. Run ac ommand in integrated terminal
2. Press Shift+Tab

Actual: JAWS reads readonly edit and you can navigate up and down the terminal output. It's super nice, because you can use the shortcut Ctrl+` to move between the code and the output of the terminal. So I can write some code, run some lines with Shift+Enter (in Python), then press Ctrl+`and focus moves back to where I was in the terminal output and I can just read down, then back to code with Ctrl+`.

Please fix that regression. I tried other flows, like that focus virtual buffer whenever a command is run, which is good, and let's keep that option, but not as good as it was in 1.82 and I find no equivalent in 1.83

VS Code version: Code - Insiders 1.83.0-insider (4268e464763087044d0c1b5bdd37ebbe683cadfa, 2023-09-29T20:42:50.630Z)
OS version: Windows_NT x64 10.0.19045
Modes:

<details>
<summary>System Info</summary>

|Item|Value|
|---|---|
|CPUs|Intel(R) Core(TM) i7-9700F CPU @ 3.00GHz (8 x 3000)|
|GPU Status|2d_canvas: enabled<br>canvas_oop_rasterization: enabled_on<br>direct_rendering_display_compositor: disabled_off_ok<br>gpu_compositing: enabled<br>multiple_raster_threads: enabled_on<br>opengl: enabled_on<br>rasterization: enabled<br>raw_draw: disabled_off_ok<br>video_decode: enabled<br>video_encode: enabled<br>vulkan: disabled_off<br>webgl: enabled<br>webgl2: enabled<br>webgpu: enabled|
|Load (avg)|undefined|
|Memory (System)|15.94GB (6.97GB free)|
|Process Argv|--crash-reporter-id 00ed479f-54df-4c3b-bb92-bdf8893179a3|
|Screen Reader|yes|
|VM|0%|
</details><details><summary>Extensions (15)</summary>

Extension|Author (truncated)|Version
---|---|---
csdevkit|ms-|0.5.127
csharp|ms-|2.4.4
vscode-dotnet-runtime|ms-|1.8.0
vscodeintellicode-csharp|ms-|0.1.26
python|ms-|2023.17.12721006
vscode-pylance|ms-|2023.9.31
vsliveshare|ms-|1.0.5883
java|red|1.23.0
intellicode-api-usage-examples|Vis|0.2.8
vscodeintellicode|Vis|1.2.30
vscode-java-debug|vsc|0.54.0
vscode-java-dependency|vsc|0.23.1
vscode-java-pack|vsc|0.25.14
vscode-java-test|vsc|0.40.0
vscode-maven|vsc|0.42.0

</details><details>
<summary>A/B Experiments</summary>

```
vsliv695:30137379
vsins829:30139715
vsliv368cf:30146710
vsreu685:30147344
python383:30185418
vspor879:30202332
vspor708:30202333
vspor363:30204092
vslsvsres303:30308271
pythontb:30258533
pythonptprofiler:30281269
vshan820:30294714
vscod805:30301674
bridge0708:30335490
bridge0723:30353136
vsaa593cf:30376535
pythonvs932:30404738
py29gd2263:30784851
vsclangdf:30492506
c4g48928:30535728
dsvsc012:30540252
pynewext54:30618038
a9j8j154:30646983
showlangstatbar:30737417
7ij38806:30687742
pythonfmttext:30716741
fixshowwlkth:30771523
showindicator:30805243
pythongtdpath:30726887
i26e3531:30792625
welcomedialog:30812478
pythonnosmt12:30779711
pythonidxpt:30768918
pythonnoceb:30776497
copilotsettingt:30808721
synctok:30821570
dsvsc013:30777762
dsvsc014:30777825
diffeditorv2:30786206
pythonmypyww:30841056
pythonmpsinfo:30842935
dsvsc015:30821418
pythontestfixtcf:30826907
pythonfb280951:30830809
pythonregdiag:30842812
pythonmypyd1:30850305
pythoncet0cf:30848248

```

</details>

**Comment by meganrogge** [comment]
Hello, we have moved away from `Shift+tab` as the default in favor of `Ctrl/Cmd+UpArrow` or `Alt+f2` to better align with how other features work around the workbench. You can assign `Shift+tab` to the action if you prefer that.

**Comment by MS-AlexB** [comment]
@meganrogge There is a change versus 1.82 around focus management. Let me give you a scenario and what it would look like in 1.82 and 1.83. Note: I will use Ctrl+Up for 1.83 and Shift+Tab for 1.82.

Scenario 1: 1.82
1. Open any file, let's say you are writing some code on line 50 of a file.
2. Press Ctrl+`, focus moves to integrated terminal
3. Press Shift+Tab, focus moves to accessible buffer
4. Use up/down to read the content, let's say you stop on line 20 of the output.
5. Press Ctrl+`, focus moves back to line 50 in the code editor.
6. Press Ctrl+`, focus moves to line 20 of the output.

Scenario 1: 1.83
1. Open any file, let's say you are writing some code on line 50 of a file.
2. Press Ctrl+`, focus moves to integrated terminal
3. Press Ctrl+Up, focus moves to accessible buffer
4. Use up/down to read the content, let's say you stop on line 20 of the output.
5. Press Ctrl+`, focus moves to the input of integrated terminal.
6. Press Ctrl+`again, focus moves to line 50 in code editor.
7. Press Ctrl+` a third time, focus moves to input of integrated terminal.

This is annoying, there is no way anymore to easily go back and forth between code and output.

**Comment by meganrogge** [comment]
Thanks for the steps. Now I understand. I just discussed this with @jooyoungseo and he prefers the approach as it currently works, so we should make this configurable.

**Comment by VSCodeTriageBot** [comment]
This feature request is now a candidate for our backlog. The community has 60 days to [upvote](https://github.com/microsoft/vscode/wiki/Issues-Triaging#up-voting-a-feature-request) the issue. If it receives 20 upvotes we will move it to our backlog. If not, we will close it. To learn more about how we handle feature requests, please see our [documentation](https://aka.ms/vscode-issue-lifecycle).

Happy Coding!

**Comment by MS-AlexB** [comment]
Thanks @meganrogge  for putting this in the backlog. Can you elaborate on the solution you have in mind to solve for this scenario? I think there are different ways to design a solution to this problem and was curious to understand more what you were thinking.

Maybe we could have a focus: terminal buffer action? I could map Ctrl+Shift+` to it and it could move my focus to the terminal buffer wherever my focus was.

So the scenario could be like:
1. Open any file, let's say you are writing some code on line 50 of a file.
2. Press Ctrl+`, focus moves to integrated terminal
3. Press Alt+F2, , focus moves to accessible buffer
4. Use up/down to read the content, let's say you stop on line 20 of the output.
5. Press Ctrl+1, focus moves back to line 50 in the code editor.
6. Press Ctrl+Shift+` (action focus: terminal buffer), focus moves to line 20 of the output.

That way Ctrl+`brings you to terminal input where you can input commands, Ctrl+Shift+` brings you to the terminal output instead - and this is the important part - wherever your focus lastw as in the output. Plus, Ctrl+Shift+` would work from anywhere, including from the code editor, which would ease frequent back and forth between code and terminal output, such as in Python flows.

**Comment by meganrogge** [comment]
I'm imagining we have a setting "preserve line" vs "last line" that determines what is focused when the accessible terminal view is opened. 

Pressing `ctrl+down` arrow will always move you to the last line in the buffer if you're not already there, so that case is covered.  Pressing `ctrl+down` arrow on the last line refocuses the terminal buffer.

The `ctrl+up`/`ctrl+down` arrow flow is how many other workbench components (extensions view input to results, debug input box to results, and more) work atm, which is why this makes sense.



Would that work for you?

**Comment by MS-AlexB** [comment]
@meganrogge maybe I misunderstood, but I believe your solution would always require at least two navigation actions to go from the code to the terminal output: Ctrl+`, then Ctrl+Up. I'd like to avoid having to do two navigation actions every time to do back and forth. Would there be a way to get to terminal output from anywhere?

**Comment by meganrogge** [comment]
You can either add a keybinding which chains those together like this:

```
{
		"key": "ctrl+3",
		"command": "runCommands",
		"when": "accessibilityModeEnabled",
		"args": {
		  "commands": [
			"workbench.action.terminal.focus",
			"workbench.action.terminal.focusAccessibleBuffer"
		  ]
		}
	},
```

The other thing you might find useful is the `terminal.integrated.focusAfterRun` setting. If you set that to `accessible-buffer` then run something in the terminal, the accessible terminal view is focused, displaying the output.

**Comment by MS-AlexB** [comment]
Wow, that shortcut works like a charm! Could we maybe add a default shortcut? Maybe Ctrl+Shift+` for that action "workbench.action.terminal.focusAccessibleBuffer"?

With that shortcut, the only thing missing would indeed be that preserve line setting.

One thing I might add on the behavior of preserve line setting:

If your focus is at the end of the buffer (Ctrl+End), and then you run a command, if preserve line setting is enabled, then focus does not move, and when you get back to the buffer, pressing Down arrow will be the command you just ran.

This would solve the annoying issue that whenever you read terminal output, you always need to first scroll back up to read down the content. For instance, if a command outputs 20 lines of text, then you need to scroll back up 20 lines to read down 20 lines. With a preserve line setting, then you can just read down, run something else, get back to buffer and read down from where you were.

**Comment by meganrogge** [comment]
`Ctrl+Shift+`` is currently assigned to creating a new terminal, so we don't want to conflict with that and break muscle memory. 

It's hard to find a keybinding that's not already in use, which is why it might be best if you assign this yourself.

**Comment by meganrogge** [comment]
@MS-AlexB that is by design. I figured when a user runs another command, they'll be more interested in that output than wherever they might have been before. This works just like a terminal does, which is what we want to model here.

**Comment by MS-AlexB** [comment]
@meganrogge but what if I run a few commands back to back? Isn't the setting supposed to preserve cursor position rather than moving it to the last command's output?

What I was hoping with this setting is
1. User runs a command
2. User reviews the output
3. User runs multiple commands
4. User can pick up where they left off and review the new output.

This setting's current implementation is skipping outputs, it's moving the cursor when it's supposed to preserve it. If I wanted the cursor to move, I would not set the preserve position setting on.

Note that preserving cursor position is a very desirable behavior in a lot of apps. For instance, if I read a Word doc, alt+tab away and come back later, I expect my cursor to stay put so I can resume reading. Same thing for web pages, emails, etc. This is what I wanted.

Imagine an app that logs things, I want to go to the output and resume reading logs. Or if I work on a Python script, run a few commands and then go review the output of those few commands.

**Comment by meganrogge** [comment]
Ok, I misunderstood your request. Thanks for the explanation.

**Comment by meganrogge** [comment]
Note that you'll have to create a new terminal for the setting to take effect. 

So, to verify:

- Set `'terminal.integrated.accessibleViewPreserveCursorPosition'` to `true`. 
- Open a terminal, enter a command, open the accessible view.
- The cursor should be on the last line
- Move the cursor up
- Run another command in the terminal
- Open the accessible view
- The cursor position should be preserved

- Set `'terminal.integrated.accessibleViewPreserveCursorPosition'` to `false`. 
- Open a terminal, enter a command, open the accessible view.
- The cursor should be on the last line
- Move the cursor up
- Run another command in the terminal
- Open the accessible view
- The cursor should be on the last line

Note that if your cursor position is at the first column and line, then it will get reset to the bottom. This is because when the editor's contents change, the cursor by default gets set to be at column & line 1. Thus, we use this condition to determine when it should be set to the bottom IE initially.

**Comment by MS-AlexB** [comment]
@meganrogge 

Your initial description of the setting on and off seems correct to me. Let me add though that if the content of terminal gets cleared, then cursor position resets to the bottom. I suspect that would already be the case without any code change, but just wanted to spec out this desired behavior.

To validate the setting behavior when you have the new implementation, test this:
1. Run "echo test1" in terminal
2. Focus accessible buffer, move cursor right before the 1 in "test1"
3. Focus terminal input, run "echo test2", then "echo test3"
4. Focus accessible buffer, cursor is still before the 1 of "test1"

For the dynamic output script that you talk about, why would the behavior of that setting off change to be preserved? Why would cursor no longer be sent to the bottom?

About the edge case of if cursor is at (1,1), then sending the cursor to the bottom is acceptable and thanks for the call out.

**Comment by meganrogge** [comment]
Thanks for the testing steps, those work based on my testing of the PR.

One other thing to note - this accessible view is meant to be shared across components and several terminals share the same one. So, we will not preserve position when one switches between terminals, as we currently don't have a mechanism to do so. In this case, it will default to the bottom. I hope that's not a big issue for you. I've noticed that generallly, JooYoung is only working in one terminal at a time.

**Comment by meganrogge** [comment]
And yeah I agree with you, we can let the setting decide in the case of dynamic output what happens. The reason I suggested otherwise was it used to work like I described.

**Comment by MS-AlexB** [comment]
Works like a charm. I typically work on one terminal at a time, will open a new issue if I encounter workflows that require a lot of switching between terminals.
